sunkit_image/granule.py

```diff
@@ -0,0 +1,298 @@
+    labels : `numpy.ndarray`
+        An array of labels with:
+
+        * 0 are granules
+        * 2 are intergranules
+        * 1 are in-between regions
+    """
+    x_size = np.shape(data)[0]
+    y_size = np.shape(data)[1]
+    data_flat = np.reshape(data, (x_size * y_size, 1))
+    labels_flat = KMeans(n_clusters=3, n_init="auto").fit(data_flat).labels_
+    labels = np.reshape(labels_flat, (x_size, y_size))
+    # Make intergranules = 0 and granules = 1.
+    group0_mean = np.mean(data[labels == 0])
+    group1_mean = np.mean(data[labels == 1])
+    group2_mean = np.mean(data[labels == 2])
+    min_index = np.argmin([group0_mean, group1_mean, group2_mean])
+    segmented_map = np.ones(labels.shape)
+    segmented_map[[labels[:, :] == min_index][0]] -= 1
+    return segmented_map
+
+
+def correlation(segment1, segment2):
+    """
+    Compute the correlation of two segmented arrays.
+
+        -1 if the agreement between two arrays is low, 0 otherwise.
+
+        Designed to be used with `segment` and `segment_kmeans` function.
+
+    Parameters
+    ----------
+    segment1 : `numpy.ndarray`
+        Main array to compare the other input array against.
+    segment2 : `numpy.ndarray`
+        Other array, i.e., data segmented using k-means clustering.
+
+    Returns
+    -------
+    [label, confidence] : `list`
+        label : `int`
+            Summarizes the confidence metric:
+            -1: if agreement is low (below 75%)
+            0: otherwise
+        confidence : `float`
+            The numeric confidence metric:
+            Between 0 and 1 where 0 if there is no agreement and 1 if completely agrees.
+    """
+    total_granules = np.count_nonzero(segment1 == 1)
+    total_intergranules = np.count_nonzero(segment1 == 0)
+    if total_granules == 0:
+        raise ValueError("No granules in `segment1`. It is possible the clustering failed.")
+    if total_intergranules == 0:
+        raise ValueError("No intergranules in `segment1`. It is possible the clustering failed.")
+    x_size = np.shape(segment1)[0]
+    y_size = np.shape(segment1)[1]
+    granule_agreement_count = 0
+    intergranule_agreement_count = 0
+    for i in range(x_size):
+        for j in range(y_size):
+            if segment1[i, j] == 1 and segment2[i, j] == 1:
+                granule_agreement_count += 1
+            elif segment1[i, j] == 0 and segment2[i, j] == 0:
+                intergranule_agreement_count += 1
+    percentage_agreement_granules = granule_agreement_count / total_granules
+    percentage_agreement_intergranules = intergranule_agreement_count / total_intergranules
+    confidence = np.mean([percentage_agreement_granules, percentage_agreement_intergranules])
+    if percentage_agreement_granules < 0.75 or percentage_agreement_intergranules < 0.75:
+        logging.info("Low agreement with K-Means clustering. Saved output has low confidence.")
+        return [-1, confidence]
+    else:
+        return [0, confidence]

```

Comment:
Why does this return -1 or 0? Why not just warn the user and return the confidence? I can't see anywhere in the code where I've missed this being used (though please tell me if I'm wrong here).

Reply:
If a user is meant to use `correlation` directly (as opposed to it being called internally by `segment`), I'd recommend that this also take in two `sunpy.Map` objects rather than just bare arrays.

Reply:
I agree. Both `segment_kmeans` and `correlation` were written by a colaborator, and I will need to check with him regarding the reasoning behind the return values of -1 and 0. Personally, I do think it might be cleaner to just print the warning and not return anything. In that case, having it called by `segment` would likley make the most sense. The original goal with these functions was simply to ensure that our method was doing better than a simple kmeans segmentation.

Reply:
I think a tool to calculate the percentage overlap between two masks is useful, especially in comparing the effects of changing some of the input parameters or the skimage_method. So I wouldn't mind seeing that remain as an available function.

Reply:
That makes sense. 
So I suppose the two possible modifications would be:
1. Use this as a check within `segemt`: I would call `segment_kmeans` within `correlation` (probably change the name to `check_against_kmeans`), and instead of returning values simply print a warning if the correlation is too low.
2. Leave this as a seperate function that could be called by the user on the final map if desired. In this case, I would change the inputs to two SunPy maps, where the first is the segmented map. I could either mae the second be a map that has been segmented using `segment_kmeans`, or I could have it be the original data, in which case I would call `segment_kmeans` within the the function (and change the name to `check_against_kmeans`).

What are your thoughts?

Reply:
I think the second option is more flexible to multiple use cases. One could compare two segment maps whether they come from the skimage and kmeans, or different skimage methods, or even two subsequent images in a time series. I think returning a dictionary with the overlap (correlation) values for each of the identified feature classes would be quick and powerful.  

I still don't like calling this `correlation` as that doesn't really capture what is being calculated as well as `overlap`
"the granules in these two segment masks overlap by 78%" make more sense (to me) than the "the two segment masks correlate at 78%"

Reply:
Great points.